From 3a85af426b51a28d9a5f4dac84c9f19af492e693 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Wed, 14 Apr 2021 11:20:36 +0300
Subject: [PATCH] [SW-0]: preboot: configure arc as privileged

Required for enabling security by LKD, this patch add changes
to set ARC0 and ARC1 as secured and privileged.

Change-Id: Iad5c99dec37b7a9936b66878fc199c051397da99
---

diff --git a/greco/Makefile b/greco/Makefile
index 2158915..4bfd3f5 100644
--- a/greco/Makefile
+++ b/greco/Makefile
@@ -1,6 +1,7 @@
 FW_SIZE := 0x4400
 PPBOOT_MAX_SIZE := 18432
 PREBOOT_SRCS_C += $(shell ls preboot/thermal_v2.c)
+PREBOOT_SRCS_C += $(shell ls $(HL_DEV)/security.c)
 
 CFLAGS += -mstrict-align
 
diff --git a/greco/security.c b/greco/security.c
new file mode 100644
index 0000000..61daddc
--- /dev/null
+++ b/greco/security.c
@@ -0,0 +1,33 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * (C) Copyright 2021 HabanaLabs, Ltd.
+ * All rights reserved.
+ *
+ */
+#include <preboot.h>
+#include <psoc_arc0_mstr_if_core_hbw_regs.h>
+#include <psoc_arc0_mstr_if_core_lbw_regs.h>
+#include <psoc_arc1_mstr_if_core_hbw_regs.h>
+#include <psoc_arc1_mstr_if_core_lbw_regs.h>
+#include "uart.h"
+
+#define EN_ARC_SEC_PRIV		0x3
+
+void preboot_secure_arc(void)
+{
+	/* enable security and privilege mode for arc0 and arc1 */
+	regmap_writel(EN_ARC_SEC_PRIV,
+		mmPSOC_ARC0_MSTR_IF_CORE_HBW_AXPROT_OVRD);
+	regmap_writel(EN_ARC_SEC_PRIV,
+		 mmPSOC_ARC0_MSTR_IF_CORE_LBW_AXPROT_OVRD);
+	regmap_writel(EN_ARC_SEC_PRIV,
+		 mmPSOC_ARC1_MSTR_IF_CORE_HBW_AXPROT_OVRD);
+	regmap_writel(EN_ARC_SEC_PRIV,
+		 mmPSOC_ARC1_MSTR_IF_CORE_LBW_AXPROT_OVRD);
+}
+
+void preboot_security_init(void)
+{
+	/* TODO add check for security strap ON */
+	preboot_secure_arc();
+}
diff --git a/greco_arc_gnu/security.c b/greco_arc_gnu/security.c
index ba92d98..4413e4b 100644
--- a/greco_arc_gnu/security.c
+++ b/greco_arc_gnu/security.c
@@ -12,8 +12,33 @@
 #include <sha256.h>
 #include <greco/greco_security.h>
 
+#include <psoc_arc0_mstr_if_core_hbw_regs.h>
+#include <psoc_arc0_mstr_if_core_lbw_regs.h>
+#include <psoc_arc1_mstr_if_core_hbw_regs.h>
+#include <psoc_arc1_mstr_if_core_lbw_regs.h>
+
+#define EN_ARC_SEC_PRIV				0x3
 #define EFUSE_SYSCFG_ENABLE_VALIDATION_MASK	BIT(10)
 
+void preboot_secure_arc(void)
+{
+	/* enable security and privilege mode for arc0 and arc1 */
+	regmap_writel(EN_ARC_SEC_PRIV,
+		mmPSOC_ARC0_MSTR_IF_CORE_HBW_AXPROT_OVRD);
+	regmap_writel(EN_ARC_SEC_PRIV,
+		 mmPSOC_ARC0_MSTR_IF_CORE_LBW_AXPROT_OVRD);
+	regmap_writel(EN_ARC_SEC_PRIV,
+		 mmPSOC_ARC1_MSTR_IF_CORE_HBW_AXPROT_OVRD);
+	regmap_writel(EN_ARC_SEC_PRIV,
+		 mmPSOC_ARC1_MSTR_IF_CORE_LBW_AXPROT_OVRD);
+}
+
+void preboot_security_init(void)
+{
+	/* TODO add check for security strap ON */
+	preboot_secure_arc();
+}
+
 static u32 sp_sram_readl(u32 offset)
 {
 	u32 reg = *((volatile u32 *) (HL_SCRATCHPAD_SRAM_BASE + offset));
